### ISSUE TITLE
Ignore <template> slide elements

### ISSUE THREAD

**Issue by hannahwoodward** [davidcetinkaya/embla-carousel#73]
I'm dynamically generating a list of slides (using AlpineJS `x-for`) before calling `EmblaCarousel`, however it seems like the `<template>` element is being picked up as a slide. My HTML structure looks like this:

```
<div class="embla">
  <ul class="embla__container">
    <template x-for="item in list"><li>...</li></template>

    <!-- dynamically generated -->
    <li class="embla__slide">...</li>
    <li class="embla__slide">...</li>
    <li class="embla__slide">...</li>
    <li class="embla__slide">...</li>
    <!-- end dynamically generated -->
  </ul>
</div>
```

I'm now manually creating this list for the time-being, but is there any way to ignore the `<template>` when creating an instance of Embla?

**Comment by davidcetinkaya** [comment]
Hello Hanna (@hannahwoodward),

Thank you for your feature request. I also appreciate the clear explanation of the issue you’re experiencing. At the time of writing this is expected behavior and I don’t think there’s a straightforward way to change it without modifying the Embla source code. The docs state:

> All immediate children of the container will be treated as slides.

I’m not familiar with Apline.js, but judging from your explanation, I’m assuming you really need the template tag where it is now to generate the slides?

I don’t know what a good solution for this would  be. Maybe a query selector to target specific slides. What are your thoughts? Any suggestions?

Kindly,
David

**Comment by davidcetinkaya** [comment]
Hello again Hannah (@hannahwoodward),

Any thoughts about my response?

Best
David

**Comment by hannahwoodward** [comment]
Hi David,

Sorry for not replying sooner!

Alpine.js is a component-based framework (similar to Vue), which helps with state management & reactivity. In this case, I'm pulling some reviews from an API via JS, then feeding the results into an Alpine component, which automatically renders the list via the `x-for` attribute (in the same DOM location). I'd usually try and do this server-side, but unfortunately that's not an option.

I did read the part on the Embla docs which said all immediate children will count as slides, but since modern browsers don't render `<template>` elements anyway, I wasn't sure if Embla should ignore them too. Not sure if relevant in anyway, but one requirement for using `x-for` is that it's used on `<template>` element.

Just tried out your hack and it seems to work - in this case I don't need to do any re-rendering after the initial API call, so happy to go with that for now.

**Comment by davidcetinkaya** [comment]
Hi Hannah (@hannahwoodward),

Thanks for the clarification. Nice to hear that the hack works. Feel free to reach out again if you at some point would like me to consider implementing this feature.

Cheers!
David

**Comment by davidcetinkaya** [comment]
Hi @hannahwoodward,

FWIW, decades later, I want to let you know that it’s possible to provide your own slides from v6 and up.

```js
const embla = EmblaCarousel({ rootNode, containerNode, slideNodes }, options)
```

👉 **Please note:**

- `rootNode` is mandatory when passing an object with element nodes instead of a single element node.
- if `containerNode` is omitted, Embla will pick the first immediate child inside the `rootNode`.
- `slideNodes` has to be an array and not a nodeList.
- It’s your responsibility to make sure that any elements inside the container that aren’t slides don’t take up any space.

Best
David

**Comment by davidcetinkaya** [comment]
@hannahwoodward I’m going to mark this issue as resolved because [this solution](https://github.com/davidcetinkaya/embla-carousel/issues/73#issuecomment-990008003) makes it possible to achieve what you want.

Please let me know if you don’t think it’s resolved.

**Comment by davidjerleke** [comment]
Hi @hannahwoodward,

This feature has now been released with [v7.1.0](https://github.com/davidjerleke/embla-carousel/releases/tag/v7.1.0) and is part of the official [options](https://www.embla-carousel.com/api/options/) API.

Best,
David
